cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -2146,6 +2147,203 @@ class TakeMetaFunction : public MetaFunction {
+                                   dst->data(), 0, column->length(), 0,
+                                   dst->mutable_data());
+    }
+  }
+  auto drop_null_filter =
+      std::make_shared<BooleanArray>(batch.num_rows(), dst, nullptr, 0, 0);
+  ARROW_ASSIGN_OR_RAISE(Datum result, Filter(Datum(batch), Datum(drop_null_filter),
+                                             FilterOptions::Defaults(), ctx));
+  return result.record_batch();
+}
+
+using ::arrow::internal::Bitmap;
+
+Result<std::shared_ptr<Table>> DropNullTable(const Table& table, ExecContext* ctx) {
+  if (table.num_rows() == 0) {
+    return Table::Make(table.schema(), table.columns(), 0);
+  }
+  const int num_columns = table.num_columns();
+  int64_t null_count = 0;
+  for (int col_index = 0; col_index < num_columns; ++col_index) {
+    const ArrayVector& chunks = table.column(col_index)->chunks();
+    for (size_t chunk_index = 0; chunk_index < chunks.size(); ++chunk_index) {
+      const auto& column_chunk = chunks[chunk_index];
+      null_count += column_chunk->null_count();
+    }
+  }
+  if (null_count == 0) {
+    return Table::Make(table.schema(), table.columns(), table.num_rows());
+  }
+  if (null_count / table.num_columns() == table.num_rows()) {
+    std::vector<std::shared_ptr<ChunkedArray>> empty_table(table.num_columns());
+    for (int i = 0; i < table.num_columns(); i++) {
+      std::shared_ptr<Array> empty_array;
+      RETURN_NOT_OK(
+          CreateEmptyArray(table.column(i)->type(), ctx->memory_pool(), &empty_array));
+      empty_table[i] = std::make_shared<ChunkedArray>(ArrayVector{empty_array});
+    }
+    return Table::Make(table.schema(), empty_table, 0);
+  }
+
+  ARROW_ASSIGN_OR_RAISE(auto dst,
+                        AllocateEmptyBitmap(table.num_rows(), ctx->memory_pool()));
+  BitUtil::SetBitsTo(dst->mutable_data(), 0, table.num_rows(), true);
+
+  for (int col_index = 0; col_index < num_columns; ++col_index) {
+    const ArrayVector& chunks = table.column(col_index)->chunks();
+    std::vector<Bitmap> bitmaps(chunks.size());
+    for (size_t chunk_index = 0; chunk_index < chunks.size(); ++chunk_index) {
+      const auto& column_chunk = chunks[chunk_index];
+      bitmaps[chunk_index] = Bitmap(column_chunk->null_bitmap_data(),
+                                    column_chunk->offset(), column_chunk->length());
+    }
+    int64_t bitmap_offset = 0;
+    ARROW_ASSIGN_OR_RAISE(auto concatenated_bitmap,
+                          AllocateEmptyBitmap(table.num_rows(), ctx->memory_pool()));
+    BitUtil::SetBitsTo(concatenated_bitmap->mutable_data(), 0, table.num_rows(), true);
+
+    for (auto bitmap : bitmaps) {
+      if (bitmap.buffer()->data()) {
+        ::arrow::internal::CopyBitmap(bitmap.buffer()->data(), bitmap.offset(),
+                                      bitmap.length(),
+                                      concatenated_bitmap->mutable_data(), bitmap_offset);
+      }
```

Comment:
This can be done in a follow-up but I wonder if this copy could be avoided by passing offsets to BitmapAnd?

Reply:
Yes It cant be done avoiding copy +1

Reply:
CopyBitmap can't be skipped for now.  BitmapAnd only has an unaligned implemnentation with respect to the offset (I don't why). So if chunk->length() % 8 != 0 BitmapAnd will fail. I would suggest to create a follow-up issue for this.